plutus-core/testlib/PlutusCore/Generators/QuickCheck/Builtin.hs

```diff
@@ -184,14 +183,18 @@ genDataFromSpine els = oneof
             ]
     ]
 
-instance Arbitrary Data where
-    arbitrary = arbitrary >>= genDataFromSpine
+instance ArbitraryBuiltin Data where
+    arbitraryBuiltin = arbitrary >>= genDataFromSpine
+
+    shrinkBuiltin (Constr i ds) = ds ++ map (Constr i) (shrinkBuiltin ds)
+    shrinkBuiltin (Map ps)      = map fst ps ++ map snd ps ++ map Map (shrinkBuiltin ps)
+    shrinkBuiltin (List ds)     = ds ++ map List (shrinkBuiltin ds)
+    shrinkBuiltin (I i)         = map I (shrinkBuiltin i)
+    shrinkBuiltin (B b)         = I 0 : map B (shrinkBuiltin b)
```

Comment:
The `I 0` is bizarre. `Constr` only shrinks to `Constr`, `I` only shrinks to `I`, and so on, but why does `B` shrink to `I`? Worth a comment.

Reply:
`Constr` doesn't only shrink to `Constr`! It's 

```haskell
shrinkBuiltin (Constr i ds) = ds ++ map (Constr i) (shrinkBuiltin ds)
```

i.e. `Constr i ds` can shrink to any `d` from `ds`.

But good point, I'll add a comment.

Reply:
Added the following comment:

```
    -- We assume that @I 0@ is the smallest 'Data' object, hence it needs to be reachable from all
    -- nodes apart from @I 0@ itself. For recursive nodes ('Constr', 'Map', and 'List') we do not
    -- need to mention @I 0@, since all of those get shrunk to smaller nodes and therefore do not
    -- contribute any "unreachability". I.e. we only need to handle the base cases, 'I' and 'B'.
    -- The clause for @I i@ doesn't require mentioning @I 0@ explicitly, since we get it through
    -- general shrinking of @i@ (unless @i@ equals @0@, as desired). Hence the only places where
    -- @I 0@ has to appear explicitly is the clause for 'B'.
```

Reply:
> We assume that @I 0@ is the smallest 'Data' object, hence it needs to be reachable from all nodes apart from @I 0@ itself

Hmm, not sure why we assume that, since `B ""` or `List []` aren't any bigger than `I 0`. And you can't reach `I 0` from `List []`.

Reply:
> Hmm, not sure why we assume that, since `B ""` or `List []` aren't any bigger than `I 0`.

Arbitrarily, just to have a final destination.

> `List []`

That's a very good point, thank you, my bad, will fix.

Reply:
> Arbitrarily, just to have a final destination.

We don't need a single final destination for shrinking to work of course, it's just to have the least interesting `Data` object possible in the end and I arbitrarily decided that `I 0` is such. So that if you see anything else, you immediately start suspecting that this is where the problem is. It's not really important, just some icing on the cake.

Reply:
Made it that:

```haskell
    -- We arbitrarily assume that @I 0@ is the smallest 'Data' object just so that anything else in
    -- a counterexample gives us a clue as to what the culprit may be. Hence @I 0@ needs to be
    -- reachable from all nodes apart from @I 0@ itself. For all nodes but 'I' we achieve this by
    -- returning @[I 0]@ if there's no other way to shrink the value, i.e. either shrinking keeps
    -- going or it's time to return the smallest object. The clause for @I i@ doesn't require
    -- mentioning @I 0@ explicitly, since we get it through general shrinking of @i@ (unless @i@
    -- equals @0@, as desired).
    shrinkBuiltin (Constr i ds) = pureIfNull (I 0) $ concat
        [ ds
        , map (Constr i) $ shrinkBuiltin ds
        , map (flip Constr ds) $ shrinkBuiltin i
        ]
    shrinkBuiltin (Map ps) = pureIfNull (I 0) $ concat
        [ map fst ps
        , map snd ps
        , map Map $ shrinkBuiltin ps
        ]
    shrinkBuiltin (List ds) = pureIfNull (I 0) $ concat
        [ ds
        , map List $ shrinkBuiltin ds
        ]
    shrinkBuiltin (B b) = pureIfNull (I 0) . map B $ shrinkBuiltin b
    shrinkBuiltin (I i) = map I $ shrinkBuiltin i
```

Gonna merge the PR once CI is green.

Reply:
^ which also added shrinking of `i` in `Constr`, 'cause why not.